### PR TITLE
New `KernelFunctionOperation <: AbstractOperation` for custom kernel functions

### LINKED ISSUE #1623: KernelFunctionOperation <: AbstractOperation to enable fast reduction of custom kernel functions
Now that we have an interface for fast on-the-fly reductions, it might make sense to implement a `KernelFunctionOperation` feature that accepts custom `kernel_function(i, j, k, grid, fields...)` so these can be reduced. (And in fact, this might also provide a replacement for `KernelComputedField`). Something like

```julia
struct KernelFunctionOperation{LX, LY, LZ, P, A, G, T, K, D} <: AbstractOperation{LX, LY, LZ, A, G, T}
    op :: K
    grid :: G
    architecture :: A
    computed_dependencies :: D
    parameters :: P
end

@inline Base.getindex(κ::KernelFunctionOperation, i, j, k) = κ.op(i, j, k, grid, κ.computed_dependencies..., κ.parameters}
@inline Base.getindex(κ::KernelFunctionOperation{LX, LY, LZ, <:Nothing}, i, j, k) where {LX, LY, LZ} = κ.op(i, j, k, grid, κ.computed_dependencies...}
```

With an appropriate `compute!` (or `compute_at!`) method for ensuring that `computed_dependencies` get computed, this might be a better / leaner interface than the one provided by `KernelComputedField`, with the added, extremely useful benefit that it can be reduced on the fly just like other `AbstractOperation`.

### PR DESCRIPTION
**Pull request by glwagner**
`KernelFunctionOperation` gives users the ability to specify custom kernel functions of `(i, j, k, grid)` as abstract operations. This is useful for a variety of purposes, such as calculating vorticity on the cubed sphere (cannot be expressed in terms of normal abstract operations) or generating random numbers in kernels within the `Field` infrastructure.

```julia
using Oceananigans.Operators: ζ₃ᶠᶠᵃ # called with signature ζ₃ᶠᶠᵃ(i, j, k, grid, u, v)

grid = model.grid
u, v, w = model.velocities

ζ_op = FunctionOperation(Face, Face, Center, ζ₃ᶠᶠᵃ, grid, computed_dependencies=(u, v))

ζ = ComputedField(ζ_op) # identical to `VerticalVorticityField`

barotropic_ζ = AveragedField(ζ_op, dims=3)
```

Resolves #1623 .

### CODE REVIEW

File: test/test_abstract_operations.jl [hunk 1]
```diff
@@ -115,6 +116,20 @@ for arch in archs
             end
         end
 
+        @testset "KernelFunctionOperations [$(typeof(arch))]" begin
+            trivial_kernel_function(i, j, k, grid) = 1
+            op = KernelFunctionOperation(trivial_kernel_function, grid)
+            @test op isa KernelFunctionOperation
+
+            less_trivial_kernel_function(i, j, k, grid, u, v) = @inbounds u[i, j, k] * v[i, j, k]
```
Comment (francispoulin) [inline test/test_abstract_operations.jl:124]: Is it worth  having a test that specifies the architecture?  I noticed there were two functions and I gather these tests only use the one that doesn't specify the architecture.

File: test/test_abstract_operations.jl [hunk 2]
```diff
@@ -115,6 +116,20 @@ for arch in archs
             end
         end
 
+        @testset "KernelFunctionOperations [$(typeof(arch))]" begin
```
Comment (francispoulin) [inline test/test_abstract_operations.jl:119]: Out of curiosity, where does the arch appear in each test?

**Review by francispoulin — COMMENTED** [review]
changes look good but some tests fail so I guess that needs to be sorted out

File: test/test_abstract_operations.jl (hunk 2 above)
Comment (glwagner) [reply test/test_abstract_operations.jl:120]: In this file there's an outer loop over `archs`: 

https://github.com/CliMA/Oceananigans.jl/blob/7094b287dad338ea2cd037264fba896499650b8b/test/test_abstract_operations.jl#L78

but not sure about other files.

**Review by ali-ramadhan — APPROVED** [review]

**Pull request merged**